libc-top-half/musl/src/env/__init_tls.c

```diff
@@ -31,6 +31,23 @@ int __init_tp(void *p)
 	if (!r) libc.can_do_threads = 1;
 	td->detach_state = DT_JOINABLE;
 	td->tid = __syscall(SYS_set_tid_address, &__thread_list_lock);
+#else
+	td->detach_state = DT_JOINABLE;
+	/*
+	 * Initialize the TID to a large value, which is hopefully
+	 * less likely to conflict with host-allocated TIDs, so that
+	 * TID-based locks can work.
+	 *
+	 * For longer term, we need a better solution.
+	 * See: https://github.com/WebAssembly/wasi-threads/issues/15
+	 *
+	 * Note:
+	 * - positive values can conflict with host-allocated TIDs.
+	 * - __tl_lock and __lockfile uses TID 0 as "unlocked".
+	 * - __lockfile relies on the fact the most significant two bits
+	 *   of TIDs are 0.
+	 */
+	td->tid = 0x3fffffff;
```

Comment:
> Initialize the TID to a large value, which is hopefully
> less likely to conflict with host-allocated TIDs

So, that assumption is correct only if the host assigns the ID incrementally, which might not be the case (host can decide the implementation).

Reply:
> > Initialize the TID to a large value, which is hopefully
> > less likely to conflict with host-allocated TIDs
> 
> So, that assumption is correct only if the host assigns the ID incrementally, which might not be the case (host can decide the implementation).

right. this is a kludge at best.

i think a real fix involves spec changes. see https://github.com/WebAssembly/wasi-threads/issues/15

Reply:
I'd suggest let's wait with this PR then until the https://github.com/WebAssembly/wasi-threads/issues/15 is resolved

Reply:
I think we can update the PR based on the discussion in https://github.com/WebAssembly/wasi-threads/issues/15 @yamt ?

Reply:
> I think we can update the PR based on the discussion in [WebAssembly/wasi-threads#15](https://github.com/WebAssembly/wasi-threads/issues/15) @yamt ?

done